File: src/connectivity/wlan/lib/common/cpp/tx_vector.cc
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <wlan/common/tx_vector.h>

namespace wlan {

//
// For HT:
// Changing CBW from 20 MHz to 40 MHz advances index by 32
// Changing GI  from 800 ns to 400 ns advances index by 64
//
//  Group   tx_vec_idx_t range    PHY   GI   CBW NSS MCS_IDX
//  0         1 -  32             HT    800  20  -   0-31
//  1        33 -  64             HT    800  40  -   0-31
//  2        65 -  96             HT    400  20  -   0-31
//  3        97 - 128             HT    400  40  -   0-31
//  4       129 - 136             ERP   -    -   -   0-7
//  5       137 - 138             DSSS  -    -   -   0-1
//  6       139 - 140             CCK   -    -   -   2-3
//
//  Note: MCS_IDX is explained in the definition of TxVector.
//
// TODO(fxbug.dev/20947) VHT will be inserted between HT and ERP.

zx_status_t TxVector::FromSupportedRate(const SupportedRate& erp_rate, TxVector* tx_vec) {
  if (tx_vec == nullptr) {
    errorf("nullptr passed to TxVector::FromSupportedRate()\n");
    ZX_DEBUG_ASSERT(false);
    return ZX_ERR_INVALID_ARGS;
  }
  *tx_vec = TxVector{
      .gi = WLAN_GI__800NS,
      .cbw = WLAN_CHANNEL_BANDWIDTH__20,
      .nss = 1,
  };

  wlan_info_phy_type_t phy;
  uint8_t mcs_idx;
  uint8_t rate_val = erp_rate.rate();

  switch (rate_val) {
    case 2:
      phy = WLAN_INFO_PHY_TYPE_DSSS;
      mcs_idx = 0;
      break;
    case 4:
      phy = WLAN_INFO_PHY_TYPE_DSSS;
      mcs_idx = 1;
      break;
    case 11:
      phy = WLAN_INFO_PHY_TYPE_CCK;
      mcs_idx = 2;
      break;
    case 22:
      phy = WLAN_INFO_PHY_TYPE_CCK;
      mcs_idx = 3;
      break;
    case 12:
      phy = WLAN_INFO_PHY_TYPE_ERP;
      mcs_idx = 0;
      break;
    case 18:
      phy = WLAN_INFO_PHY_TYPE_ERP;
      mcs_idx = 1;
      break;
    case 24:
      phy = WLAN_INFO_PHY_TYPE_ERP;
      mcs_idx = 2;
      break;
    case 36:
      phy = WLAN_INFO_PHY_TYPE_ERP;
      mcs_idx = 3;
      break;
    case 48:
      phy = WLAN_INFO_PHY_TYPE_ERP;
      mcs_idx = 4;
      break;
    case 72:
      phy = WLAN_INFO_PHY_TYPE_ERP;
      mcs_idx = 5;
      break;
    case 96:
      phy = WLAN_INFO_PHY_TYPE_ERP;
      mcs_idx = 6;
      break;
    case 108:
      phy = WLAN_INFO_PHY_TYPE_ERP;
      mcs_idx = 7;
      break;
    default:
      errorf("Invalid rate %u * 0.5 Mbps for 802.11a/b/g.\n", rate_val);
      ZX_DEBUG_ASSERT(false);
      return ZX_ERR_INVALID_ARGS;
  }
  tx_vec->phy = phy;
  tx_vec->mcs_idx = mcs_idx;
  return ZX_OK;
}

// Inverse of FromSupportedRate above, supports ERP rates but not CCK rates.
std::optional<SupportedRate> TxVectorIdxToErpRate(tx_vec_idx_t idx) {
  if (idx < kErpStartIdx || idx >= kErpStartIdx + kErpNumTxVector) {
    return {};
  }
  constexpr uint16_t erp_rate_list[] = {12, 18, 34, 36, 48, 72, 96, 108};
  return SupportedRate(erp_rate_list[idx - kErpStartIdx]);
}

bool IsTxVecIdxValid(tx_vec_idx_t idx) { return kInvalidTxVectorIdx < idx && idx <= kMaxValidIdx; }

wlan_info_phy_type_t TxVecIdxToPhy(tx_vec_idx_t idx) {
  if (idx < kHtStartIdx + kHtNumTxVector) {
    return WLAN_INFO_PHY_TYPE_HT;
  } else if (idx < kErpStartIdx + kErpNumTxVector) {
    return WLAN_INFO_PHY_TYPE_ERP;
  } else if (idx < kDsssCckStartIdx + kDsssCckNumTxVector) {
    return idx - kDsssCckStartIdx < 2 ? WLAN_INFO_PHY_TYPE_DSSS : WLAN_INFO_PHY_TYPE_CCK;
  }
  // caller will always call IsTxVecIdxValie() so that this is never reached.
  ZX_DEBUG_ASSERT(false);
  return WLAN_INFO_PHY_TYPE_HT;
}

zx_status_t TxVector::FromIdx(tx_vec_idx_t idx, TxVector* tx_vec) {
  if (!IsTxVecIdxValid(idx)) {
    errorf("Invalid idx for TxVector::FromIdx(): %u\n", idx);
    ZX_DEBUG_ASSERT(false);
    return ZX_ERR_INVALID_ARGS;
  }
  if (tx_vec == nullptr) {
    errorf("nullptr for TxVector::FromIdx()\n");
    ZX_DEBUG_ASSERT(false);
    return ZX_ERR_INVALID_ARGS;
  }
  wlan_info_phy_type_t phy = TxVecIdxToPhy(idx);
  switch (phy) {
    case WLAN_INFO_PHY_TYPE_HT: {
      uint8_t group_idx = (idx - kHtStartIdx) / kHtNumMcs;
      wlan_gi_t gi = ((group_idx / kHtNumCbw) % kHtNumGi == 1 ? WLAN_GI__400NS : WLAN_GI__800NS);
      wlan_channel_bandwidth_t cbw =
          (group_idx % kHtNumCbw == 0 ? WLAN_CHANNEL_BANDWIDTH__20 : WLAN_CHANNEL_BANDWIDTH__40);
      uint8_t mcs_idx = (idx - kHtStartIdx) % kHtNumMcs;

      *tx_vec = TxVector{
          .phy = phy,
          .gi = gi,
          .cbw = cbw,
          .nss = static_cast<uint8_t>(1 + mcs_idx / kHtNumUniqueMcs),
          .mcs_idx = mcs_idx,
      };
      break;
    }
    case WLAN_INFO_PHY_TYPE_ERP:
      *tx_vec = TxVector{
          .phy = phy,
          .gi = WLAN_GI__800NS,
          .cbw = WLAN_CHANNEL_BANDWIDTH__20,
          .nss = 1,
          .mcs_idx = static_cast<uint8_t>(idx - kErpStartIdx),
      };
      break;
    case WLAN_INFO_PHY_TYPE_DSSS:
    case WLAN_INFO_PHY_TYPE_CCK:
      *tx_vec = TxVector{
          .phy = phy,
          .gi = WLAN_GI__800NS,
          .cbw = WLAN_CHANNEL_BANDWIDTH__20,
          .nss = 1,
          .mcs_idx = static_cast<uint8_t>(idx - kDsssCckStartIdx),
      };
      break;
    default:
      // Not reachable.
      ZX_DEBUG_ASSERT(false);
      break;
  }
  return ZX_OK;
}

bool TxVector::IsValid() const {
  if (!(phy == WLAN_INFO_PHY_TYPE_CCK || phy == WLAN_INFO_PHY_TYPE_DSSS ||
        phy == WLAN_INFO_PHY_TYPE_ERP || phy == WLAN_INFO_PHY_TYPE_HT)) {
    return false;
  }
  switch (phy) {
    case WLAN_INFO_PHY_TYPE_DSSS:
      return mcs_idx == 0 || mcs_idx == 1;
    case WLAN_INFO_PHY_TYPE_CCK:
      return mcs_idx == 2 || mcs_idx == 3;
    case WLAN_INFO_PHY_TYPE_HT:
      if (!(gi == WLAN_GI__800NS || gi == WLAN_GI__400NS)) {
        return false;
      }
      if (!(cbw == WLAN_CHANNEL_BANDWIDTH__20 || cbw == WLAN_CHANNEL_BANDWIDTH__40 ||
            cbw == WLAN_CHANNEL_BANDWIDTH__40ABOVE || cbw == WLAN_CHANNEL_BANDWIDTH__40BELOW)) {
        return false;
      }
      return 0 <= mcs_idx && mcs_idx < kHtNumMcs;
    case WLAN_INFO_PHY_TYPE_ERP:
      return 0 <= mcs_idx && mcs_idx < kErpNumTxVector;
    case WLAN_INFO_PHY_TYPE_VHT:
      // fall through
      // TODO(fxbug.dev/28964): GI 800ns, 400ns or 200ns, BW any, MCS 0-9
    default:
      return false;
  }
}

zx_status_t TxVector::ToIdx(tx_vec_idx_t* idx) const {
  if (!IsValid()) {
    return ZX_ERR_INVALID_ARGS;
  }
  switch (phy) {
    case WLAN_INFO_PHY_TYPE_HT: {
      uint8_t group_idx = 0;
      if (gi == WLAN_GI__400NS) {
        group_idx = kHtNumCbw;
      }
      if (cbw == WLAN_CHANNEL_BANDWIDTH__40 || cbw == WLAN_CHANNEL_BANDWIDTH__40ABOVE ||
          cbw == WLAN_CHANNEL_BANDWIDTH__40BELOW) {
        group_idx++;
      }

      *idx = kHtStartIdx + (group_idx * kHtNumMcs) + mcs_idx;
      break;
    }
    case WLAN_INFO_PHY_TYPE_ERP:
      *idx = kErpStartIdx + mcs_idx;
      break;
    case WLAN_INFO_PHY_TYPE_CCK:
    case WLAN_INFO_PHY_TYPE_DSSS:
      *idx = kDsssCckStartIdx + mcs_idx;
      break;
    case WLAN_INFO_PHY_TYPE_VHT:
      // fall-through, will never reach because TxVector is always valid.
      // TODO(fxbug.dev/28964)
    default:
      break;
  }
  return ZX_OK;
}

bool operator==(const TxVector& lhs, const TxVector& rhs) {
  if (lhs.phy != rhs.phy || lhs.mcs_idx != rhs.mcs_idx) {
    return false;
  }
  switch (lhs.phy) {
    case WLAN_INFO_PHY_TYPE_HT:
      return lhs.gi == rhs.gi && lhs.cbw == rhs.cbw;
    case WLAN_INFO_PHY_TYPE_ERP:
    case WLAN_INFO_PHY_TYPE_CCK:
    case WLAN_INFO_PHY_TYPE_DSSS:
      return true;
    default:
      return false;
  }
}

bool operator!=(const TxVector& lhs, const TxVector& rhs) { return !(lhs == rhs); }

bool IsEqualExceptMcs(const ::wlan::TxVector& lhs, const ::wlan::TxVector& rhs) {
  ::wlan::TxVector temp = lhs;
  temp.mcs_idx = rhs.mcs_idx;  // Make mcs_idx equal so that we only compare other fields.
  return rhs == temp;
}
}  // namespace wlan

File: src/connectivity/wlan/lib/common/rust/src/tx_vector.rs
// Copyright 2021 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use {
    crate::ie::SupportedRate,
    anyhow::{bail, Error},
    banjo_fuchsia_hardware_wlan_info as hw_wlan_info,
    banjo_fuchsia_hardware_wlan_mac as hw_wlan_mac, banjo_fuchsia_wlan_common as banjo_common,
};

pub const HT_NUM_MCS: u8 = 32; // Only support MCS 0-31
pub const HT_NUM_UNIQUE_MCS: u8 = 8;
pub const ERP_NUM_TX_VECTOR: u8 = 8;

const INVALID_TX_VECTOR_IDX: u16 = hw_wlan_mac::WLAN_TX_VECTOR_IDX_INVALID;

const HT_NUM_GI: u8 = 2;
const HT_NUM_CBW: u8 = 2;
const HT_NUM_TX_VECTOR: u8 = HT_NUM_GI * HT_NUM_CBW * HT_NUM_MCS;

const DSSS_CCK_NUM_TX_VECTOR: u8 = 4;

pub const START_IDX: u16 = 1 + INVALID_TX_VECTOR_IDX;
pub const HT_START_IDX: u16 = START_IDX;
pub const ERP_START_IDX: u16 = HT_START_IDX + HT_NUM_TX_VECTOR as u16;
pub const DSSS_CCK_START_IDX: u16 = ERP_START_IDX + ERP_NUM_TX_VECTOR as u16;
pub const MAX_VALID_IDX: u16 = DSSS_CCK_START_IDX + DSSS_CCK_NUM_TX_VECTOR as u16 - 1;

// Notes about HT:
// Changing CBW (channel bandwidth) from 20 MHz to 40 MHz advances index by 32
// Changing GI (gap interval) from 800 ns to 400 ns advances index by 64
//
//  Group   tx_vec_idx_t range    PHY   GI   CBW NSS MCS_IDX
//  0         1 -  32             HT    800  20  -   0-31
//  1        33 -  64             HT    800  40  -   0-31
//  2        65 -  96             HT    400  20  -   0-31
//  3        97 - 128             HT    400  40  -   0-31
//  4       129 - 136             ERP   -    -   -   0-7
//  5       137 - 138             DSSS  -    -   -   0-1
//  6       139 - 140             CCK   -    -   -   2-3
//
// TODO(fxbug.dev/20947) VHT will be inserted between HT and ERP.

#[derive(PartialEq, Debug)]
/// Encapsulates parameters for transmitting a packet over a PHY.
///
/// MCS index is defined in
/// * HT: IEEE 802.11-2016 Table 19-27
/// * VHT: IEEE 802.11-2016 Table 21-30
///
/// We extend the definition of MCS index beyond IEEE 802.11-2016 as follows:
/// * For ERP/ERP-OFDM (WlanPhyType::ERP):
///     * 0: BPSK,   1/2 -> Data rate  6 Mbps
///     * 1: BPSK,   3/4 -> Data rate  9 Mbps
///     * 2: QPSK,   1/2 -> Data rate 12 Mbps
///     * 3: QPSK,   3/4 -> Data rate 18 Mbps
///     * 4: 16-QAM, 1/2 -> Data rate 24 Mbps
///     * 5: 16-QAM, 3/4 -> Data rate 36 Mbps
///     * 6: 64-QAM, 2/3 -> Data rate 48 Mbps
///     * 7: 64-QAM, 3/4 -> Data rate 54 Mbps
/// * For DSSS, HR/DSSS, and ERP-DSSS/CCK (WlanPhyType::DSSS and WlanPhyType::CCK):
///     * 0:  2 -> 1   Mbps DSSS
///     * 1:  4 -> 2   Mbps DSSS
///     * 2: 11 -> 5.5 Mbps CCK
///     * 3: 22 -> 11  Mbps CCK
pub struct TxVector {
    phy: hw_wlan_info::WlanPhyType,
    gi: hw_wlan_info::WlanGi,
    cbw: banjo_common::ChannelBandwidth,
    nss: u8, // Number of spatial streams for VHT and beyond.
    // For HT,  see IEEE 802.11-2016 Table 19-27
    // For VHT, see IEEE 802.11-2016 Table 21-30
    // For ERP, see comment above (this is a Fuchsia extension)
    mcs_idx: u8,
}

impl TxVector {
    pub fn new(
        phy: hw_wlan_info::WlanPhyType,
        gi: hw_wlan_info::WlanGi,
        cbw: banjo_common::ChannelBandwidth,
        mcs_idx: u8,
    ) -> Result<Self, Error> {
        let supported_mcs = match phy {
            hw_wlan_info::WlanPhyType::DSSS => mcs_idx == 0 || mcs_idx == 1,
            hw_wlan_info::WlanPhyType::CCK => mcs_idx == 2 || mcs_idx == 3,
            hw_wlan_info::WlanPhyType::HT => {
                match gi {
                    hw_wlan_info::WlanGi::G_800NS | hw_wlan_info::WlanGi::G_400NS => (),
                    other => bail!("Unsupported GI for HT PHY: {:?}", other),
                }
                match cbw {
                    banjo_common::ChannelBandwidth::CBW20
                    | banjo_common::ChannelBandwidth::CBW40
                    | banjo_common::ChannelBandwidth::CBW40BELOW => (),
                    other => bail!("Unsupported CBW for HT PHY: {:?}", other),
                }
                mcs_idx < HT_NUM_MCS
            }
            hw_wlan_info::WlanPhyType::ERP => mcs_idx < ERP_NUM_TX_VECTOR,
            other => bail!("Unsupported phy type: {:?}", other),
        };
        if supported_mcs {
            let nss = match phy {
                hw_wlan_info::WlanPhyType::HT => 1 + mcs_idx / HT_NUM_UNIQUE_MCS,
                // TODO(fxbug.dev/20947): Support VHT NSS
                _ => 1,
            };
            Ok(Self { phy, gi, cbw, nss, mcs_idx })
        } else {
            bail!("Unsupported MCS {:?} for phy type {:?}", mcs_idx, phy);
        }
    }

    pub fn phy(&self) -> hw_wlan_info::WlanPhyType {
        self.phy
    }

    pub fn from_supported_rate(erp_rate: &SupportedRate) -> Result<Self, Error> {
        let (phy, mcs_idx) = match erp_rate.rate() {
            2 => (hw_wlan_info::WlanPhyType::DSSS, 0),
            4 => (hw_wlan_info::WlanPhyType::DSSS, 1),
            11 => (hw_wlan_info::WlanPhyType::CCK, 2),
            22 => (hw_wlan_info::WlanPhyType::CCK, 3),
            12 => (hw_wlan_info::WlanPhyType::ERP, 0),
            18 => (hw_wlan_info::WlanPhyType::ERP, 1),
            24 => (hw_wlan_info::WlanPhyType::ERP, 2),
            36 => (hw_wlan_info::WlanPhyType::ERP, 3),
            48 => (hw_wlan_info::WlanPhyType::ERP, 4),
            72 => (hw_wlan_info::WlanPhyType::ERP, 5),
            96 => (hw_wlan_info::WlanPhyType::ERP, 6),
            108 => (hw_wlan_info::WlanPhyType::ERP, 7),
            other_rate => {
                bail!("Invalid rate {} * 0.5 Mbps for 802.11a/b/g.", other_rate);
            }
        };
        Self::new(
            phy,
            hw_wlan_info::WlanGi::G_800NS,
            banjo_common::ChannelBandwidth::CBW20,
            mcs_idx,
        )
    }

    // We guarantee safety of the unwraps in the following two functions by testing all TxVecIdx
    // values exhaustively.

    pub fn from_idx(idx: TxVecIdx) -> Self {
        let phy = idx.to_phy();
        match phy {
            hw_wlan_info::WlanPhyType::HT => {
                let group_idx = (*idx - HT_START_IDX) / HT_NUM_MCS as u16;
                let gi = match (group_idx / HT_NUM_CBW as u16) % HT_NUM_GI as u16 {
                    1 => hw_wlan_info::WlanGi::G_400NS,
                    _ => hw_wlan_info::WlanGi::G_800NS,
                };
                let cbw = match group_idx % HT_NUM_CBW as u16 {
                    0 => banjo_common::ChannelBandwidth::CBW20,
                    _ => banjo_common::ChannelBandwidth::CBW40,
                };
                let mcs_idx = ((*idx - HT_START_IDX) % HT_NUM_MCS as u16) as u8;
                Self::new(phy, gi, cbw, mcs_idx).unwrap()
            }
            hw_wlan_info::WlanPhyType::ERP => Self::new(
                phy,
                hw_wlan_info::WlanGi::G_800NS,
                banjo_common::ChannelBandwidth::CBW20,
                (*idx - ERP_START_IDX) as u8,
            )
            .unwrap(),
            hw_wlan_info::WlanPhyType::DSSS | hw_wlan_info::WlanPhyType::CCK => Self::new(
                phy,
                hw_wlan_info::WlanGi::G_800NS,
                banjo_common::ChannelBandwidth::CBW20,
                (*idx - DSSS_CCK_START_IDX) as u8,
            )
            .unwrap(),
            _ => unreachable!(),
        }
    }

    pub fn to_idx(&self) -> TxVecIdx {
        match self.phy {
            hw_wlan_info::WlanPhyType::HT => {
                let group_idx = match self.gi {
                    hw_wlan_info::WlanGi::G_400NS => HT_NUM_CBW as u16,
                    _ => 0,
                } + match self.cbw {
                    banjo_common::ChannelBandwidth::CBW40
                    | banjo_common::ChannelBandwidth::CBW40BELOW => 1,
                    _ => 0,
                };
                TxVecIdx::new(HT_START_IDX + group_idx * HT_NUM_MCS as u16 + self.mcs_idx as u16)
                    .unwrap()
            }
            hw_wlan_info::WlanPhyType::ERP => {
                TxVecIdx::new(ERP_START_IDX + self.mcs_idx as u16).unwrap()
            }
            hw_wlan_info::WlanPhyType::CCK | hw_wlan_info::WlanPhyType::DSSS => {
                TxVecIdx::new(DSSS_CCK_START_IDX + self.mcs_idx as u16).unwrap()
            }
            _ => unreachable!(),
        }
    }
}

#[derive(Hash, PartialEq, Eq, Debug, Copy, Clone, Ord, PartialOrd)]
pub struct TxVecIdx(u16);
impl std::ops::Deref for TxVecIdx {
    type Target = u16;
    fn deref(&self) -> &u16 {
        &self.0
    }
}

impl TxVecIdx {
    pub fn new(value: u16) -> Option<Self> {
        if INVALID_TX_VECTOR_IDX < value && value <= MAX_VALID_IDX {
            Some(Self(value))
        } else {
            None
        }
    }

    // TODO(fxbug.dev/82520): Add a const fn new when it's a stable feature.

    pub fn to_erp_rate(&self) -> Option<SupportedRate> {
        const ERP_RATE_LIST: [u8; ERP_NUM_TX_VECTOR as usize] = [12, 18, 24, 36, 48, 72, 96, 108];
        if self.is_erp() {
            Some(SupportedRate(ERP_RATE_LIST[(self.0 - ERP_START_IDX) as usize]))
        } else {
            None
        }
    }

    pub fn to_phy(&self) -> hw_wlan_info::WlanPhyType {
        match self.0 {
            idx if idx < HT_START_IDX + HT_NUM_TX_VECTOR as u16 => hw_wlan_info::WlanPhyType::HT,
            idx if idx < ERP_START_IDX + ERP_NUM_TX_VECTOR as u16 => hw_wlan_info::WlanPhyType::ERP,
            idx if idx < DSSS_CCK_START_IDX + 2 => hw_wlan_info::WlanPhyType::DSSS,
            idx if idx < DSSS_CCK_START_IDX + DSSS_CCK_NUM_TX_VECTOR as u16 => {
                hw_wlan_info::WlanPhyType::CCK
            }
            // This panic is unreachable for any TxVecIdx constructed with TxVecIdx::new.
            // Verified by exhaustive test cases.
            _ => panic!("TxVecIdx has invalid value"),
        }
    }

    pub fn is_ht(&self) -> bool {
        HT_START_IDX <= self.0 && self.0 < HT_START_IDX + HT_NUM_TX_VECTOR as u16
    }

    pub fn is_erp(&self) -> bool {
        ERP_START_IDX <= self.0 && self.0 < ERP_START_IDX + ERP_NUM_TX_VECTOR as u16
    }
}

impl std::fmt::Display for TxVecIdx {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let tx_vector = TxVector::from_idx(*self);
        write!(f, "TxVecIdx {:3}: {:?}", self.0, tx_vector)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn valid_tx_vector_idxs() {
        for idx in INVALID_TX_VECTOR_IDX + 1..=MAX_VALID_IDX {
            let idx = TxVecIdx::new(idx).expect("Could not make TxVecIdx from valid index");
            idx.to_phy(); // Shouldn't panic for any value.
        }
        assert!(
            TxVecIdx::new(INVALID_TX_VECTOR_IDX).is_none(),
            "Should not be able to construct invalid tx vector idx"
        );
        assert!(
            TxVecIdx::new(MAX_VALID_IDX + 1).is_none(),
            "Should not be able to construct invalid tx vector idx"
        );
    }

    #[test]
    fn erp_rates() {
        for idx in INVALID_TX_VECTOR_IDX + 1..=MAX_VALID_IDX {
            let idx = TxVecIdx::new(idx).expect("Could not make TxVecIdx from valid index");
            assert_eq!(idx.is_erp(), idx.to_erp_rate().is_some());
        }
    }

    #[test]
    fn phy_types() {
        for idx in INVALID_TX_VECTOR_IDX + 1..=MAX_VALID_IDX {
            let idx = TxVecIdx::new(idx).expect("Could not make TxVecIdx from valid index");
            if idx.is_erp() {
                assert_eq!(idx.to_phy(), hw_wlan_info::WlanPhyType::ERP);
            } else if idx.is_ht() {
                assert_eq!(idx.to_phy(), hw_wlan_info::WlanPhyType::HT);
            } else {
                assert!(
                    idx.to_phy() == hw_wlan_info::WlanPhyType::DSSS
                        || idx.to_phy() == hw_wlan_info::WlanPhyType::CCK
                );
            }
        }
    }

    #[test]
    fn to_and_from_idx() {
        for idx in INVALID_TX_VECTOR_IDX + 1..=MAX_VALID_IDX {
            let idx = TxVecIdx::new(idx).expect("Could not make TxVecIdx from valid index");
            let tx_vector = TxVector::from_idx(idx);
            assert_eq!(idx, tx_vector.to_idx());
        }
    }

    #[test]
    fn ht_and_erp_phy_types() {
        for idx in INVALID_TX_VECTOR_IDX + 1..=MAX_VALID_IDX {
            let idx = TxVecIdx::new(idx).expect("Could not make TxVecIdx from valid index");
            let tx_vector = TxVector::from_idx(idx);
            if idx.is_erp() {
                assert_eq!(tx_vector.phy(), hw_wlan_info::WlanPhyType::ERP);
            } else if idx.is_ht() {
                assert_eq!(tx_vector.phy(), hw_wlan_info::WlanPhyType::HT);
            }
        }
    }

    #[test]
    fn from_erp_rates() {
        for idx in INVALID_TX_VECTOR_IDX + 1..=MAX_VALID_IDX {
            let idx = TxVecIdx::new(idx).expect("Could not make TxVecIdx from valid index");
            if idx.is_erp() {
                let erp_rate = idx.to_erp_rate().unwrap();
                let tx_vector = TxVector::from_supported_rate(&erp_rate)
                    .expect("Could not make TxVector from ERP rate.");
                assert_eq!(idx, tx_vector.to_idx());
            }
        }
    }
}
